samples/GalleryApp/ObjectDetectorPage.xaml.cs

```diff
@@ -0,0 +1,494 @@
+                        await m_processedBitmapSource.SetBitmapAsync(bitmap);
+                    }
+
+                    // Retrieve and filter results if requested
+                    IReadOnlyList<ObjectDetectorResult> objectDetections = m_binding.DetectedObjects;
+                    if (m_objectKinds?.Count > 0)
+                    {
+                        objectDetections = objectDetections.Where(det => m_objectKinds.Contains(det.Kind)).ToList();
+                    }
+
+                    // Update displayed results
+                    m_bboxRenderer.Render(objectDetections);
+
+                    // Update the displayed performance text
+                    UIPerfTextBlock.Text = $"bind: {m_bindTime.ToString("F2")}ms, eval: {m_evalTime.ToString("F2")}ms";
+                }
+                catch (TaskCanceledException)
+                {
+                    // no-op: we expect this exception when we change media sources
+                    // and can safely ignore/continue
+                }
+                catch (Exception ex)
+                {
+                    NotifyUser($"Exception while rendering results: {ex.Message}");
+                }
+            });
+        }
+
+        /// <summary>
+        /// Configure an IFrameSource from a StorageFile or MediaCapture instance
+        /// </summary>
+        /// <param name="source"></param>
+        /// <returns></returns>
+        private async Task ConfigureFrameSourceAsync(object source)
+        {
+            await m_lock.WaitAsync();
+            {
+                // Reset bitmap rendering component
+                UIProcessedPreview.Source = null;
+                m_processedBitmapSource = new SoftwareBitmapSource();
+                UIProcessedPreview.Source = m_processedBitmapSource;
+
+                // Clean up previous frame source
+                if (m_frameSource != null)
+                {
+                    m_frameSource.FrameArrived -= FrameSource_FrameAvailable;
+                    var disposableFrameSource = m_frameSource as IDisposable;
+                    if (disposableFrameSource != null)
+                    {
+                        // Lock disposal based on frame source consumers
+                        disposableFrameSource.Dispose();
+                    }
+                }
+
+                // Create new frame source and rgister a callback if the source fails along the way
+                m_frameSource = await FrameSourceFactory.CreateFrameSourceAsync(source, (sender, message) =>
+                {
+                    NotifyUser(message);
+                });
+
+                // TODO: Workaround for a bug in ObjectDetectorBinding when binding consecutively VideoFrames with Direct3DSurface and SoftwareBitmap
+                m_binding = await m_skill.CreateSkillBindingAsync() as ObjectDetectorBinding;
+            }
+            m_lock.Release();
+
+            // If we obtained a valid frame source, start it
```

Comment:
>If we obtained a valid frame source, start it [](start = 15, length = 45)

otherwise will there be an error message? What is the practical case when a valid frame isn't available - bug in the camera driver resulting in camera not functioning?

Reply:
Yes, there will be error message from the NotifyUser(message) a few line above. 
We don't want other app developer to introduce a new type of frame source (which is not Windows.Storage.StorageFile or a Windows.Devices.Enumeration.DeviceInformation) to the program. And this will be cause the a null frame source.

---
In reply to: [297793317](https://github.com/microsoft/WindowsVisionSkillsPreview/pull/39#discussion_r297793317) [](ancestors = 297793317)

Reply:
got it, it might be worth evaluating error use cases that we won't to surface error messages for; in this case if a new input type is added the error panel will only display loading source without any indication of error.

---
In reply to: [299617755](https://github.com/microsoft/WindowsVisionSkillsPreview/pull/39#discussion_r299617755) [](ancestors = 299617755,297793317)